python/rascaline/rascaline/utils/clebsch_gordan/_cg_cache.py

```diff
@@ -18,55 +20,174 @@
+    ``cg_backend="python-sparse"`` or ``cg_backend="mops"`` (i.e. sparse with MOPS)::
+        - samples: `(m1, m2, mu)`, where `m1` and `m2` are the m component
+          values for the two arrays being combined and `mu` is the m component
+          value for the resulting array.
+        - components: `[]`, i.e. no components axis.
+        - properties: `property`, i.e. a dummy property.
+
+
+    :param lambda_max: maximum lambda value to compute CG coefficients for.
+    :param sparse: whether to store the CG coefficients in sparse format.
+    :param use_mops: whether to store the CG coefficients in MOPS sparse format. This is
+        recommended as the default for sparse accumulation, but can only be used if Mops
+        is installed.
+    :param use_torch: whether torch tensor or numpy arrays should be used for the cg
+        coeffs
+
+    :returns: :py:class:`TensorMap` of the Clebsch-Gordan coefficients.
     """
-    Class for computing Clebsch-Gordan coefficients for real spherical
-    harmonics.
+    # Parse the various boolean options for array and CG combination backends
+    sparse, use_mops, use_torch = _parse_backend_options(sparse, use_mops, use_torch)
+
+    # Build some 'like' arrays for the backend dispatch
+    if use_torch:
+        complex_like = torch.empty(0, dtype=torch.complex128)
+        double_like = torch.empty(0, dtype=torch.double)
+        if isinstance(Labels, torch.ScriptClass):
+            labels_values_like = torch.empty(0, dtype=torch.double)
+        else:
+            labels_values_like = np.empty(0, dtype=np.double)
+    else:
+        complex_like = np.empty(0, dtype=np.complex128)
+        double_like = np.empty(0, dtype=np.double)
+        labels_values_like = np.empty(0, dtype=np.double)
+
+    # Calculate the CG coefficients, stored as a dict of dense arrays. This is the
+    # starting point for the conversion to a TensorMap of different formats depending on
+    # the backend options.
+    cg_coeff_dict = _build_dense_cg_coeff_dict(
+        lambda_max, complex_like, double_like, labels_values_like
+    )
+
+    # Build the CG cache depending on whether the CG backend is sparse or dense. The
+    # dispatching of the arrays backends are accounted for by `double_like` and
+    # `labels_values_like`.
+    if sparse:
+        return _cg_coeff_dict_to_tensormap_sparse(
+            cg_coeff_dict, double_like, labels_values_like
+        )
 
-    Stores the coefficients in a dictionary in the `self.coeffs` attribute,
-    which is built at initialization. There are 3 current use cases for the
-    format of these coefficients. By default, sparse accumulation of products is
-    performed, whether or not Mops is installed.
+    return _cg_coeff_dict_to_tensormap_dense(
+        cg_coeff_dict, double_like, labels_values_like
+    )
 
-    Case 1: standard sparse format.
 
-    Each dictionary entry is a dictionary with entries for each (m1, m2, mu)
-    combination.
+def _parse_backend_options(
```

Comment:
if we don't have `use_mops` at this level, I think we can also get rid of this function